File: TypeScriptSource/Container/List.ts
//reference
/// <reference path="ListNode.ts" />

//namespace declaration
namespace Nolix.List {

    //cass
    export class List<E> {

        //attribute
        private elementCount: number = 0;

        //optional attributes
        private beginNode: ListNode<E>;
        private endNode: ListNode<E>;

        //method
        public addAtBegin(element: E): List<E> {

            const previousNode = new ListNode<E>(element);

            if (this.isEmpty) {
                this.beginNode = previousNode;
                this.endNode = previousNode;
            }
            else {
                previousNode.setNextNode(this.beginNode);
                this.beginNode = previousNode;
            }

            this.elementCount++;

            return this;
        }

        //method
        public addAtEnd(element: E): List<E> {

            const nextNode = new ListNode<E>(element);

            if (this.isEmpty) {
                this.beginNode = nextNode;
                this.endNode = nextNode;
            }
            else {
                this.endNode.setNextNode(nextNode);
                this.endNode = nextNode;
            }

            this.elementCount++;

            return this;
        }

        //method
        public containsAnyElement() {
            return (this.getElementCount() > 0);
        }

        //method
        public getElementCount(): number {
            return this.elementCount;
        }

        //method
        public getRefFirstElement(): E {
            
            this.supposeIsNotEmpty();

            return this.beginNode.getRefElement();
        }

        //method
        public getRefLastElement(): E {

            this.supposeIsNotEmpty();

            return this.endNode.getRefElement();
        }

        //method
        public isEmpty() {
            return (this.getElementCount() == 0);
        }

        //method
        private supposeIsNotEmpty() {
            if (this.isEmpty) {
                throw new Error("The current list is empty.");
            }
        }
    }
}

File: Source/ch/nolix/common/independentContainers/List.java
//package declaration
package ch.nolix.common.independentContainers;

//Java import
import java.util.Iterator;

//class
public final class List<E> implements Iterable<E> {
	
	//attribute
	private int elementCount = 0;
	
	//optional attributes
	private ListNode<E> beginNode;
	private ListNode<E> endNode;
	
	//constructor
	public List() {}
	
	//constructor
	public List(final E element) {
		addAtBegin(element);
	}
	
	//constructor
	public List(final E[] elements) {
		for (final var e : elements) {
			addAtBegin(e);
		}
	}
	
	//method
	public void addAtBegin(final E element) {
		
		final ListNode<E> node = new ListNode<>(element);
		
		if (isEmpty()) {
			beginNode = node;
			endNode = node;
		}
		else {
			node.setNextNode(beginNode);
			beginNode = node;
		}
		
		elementCount++;
	}
	
	//method
	public void addAtEnd(final E element) {
		
		final var node = new ListNode<E>(element);
		
		if (isEmpty()) {
			beginNode = node;
			endNode = node;
		}
		else {
			endNode.setNextNode(node);
			endNode = node;
		}
		
		elementCount++;
	}
	
	//method
	public void clear() {
		beginNode = null;
		endNode = null;
		elementCount = 0;
	}
	
	//method
	public int getElementCount() {
		return elementCount;
	}
	
	//method
	public E getRefFirst() {
		
		supposeIsNotEmpty();
		
		return beginNode.getRefElement();
	}

	//method
	public boolean isEmpty() {
		return (beginNode == null);
	}
	
	//method
	/**
	 * @return a new {@link Iterator} for the current {@link List}.
	 */
	@Override
	public Iterator<E> iterator() {
		return new ListIterator<>(beginNode);
	}
	
	//method
	public void removeFirst() {
		
		supposeIsNotEmpty();
		
		if (!beginNode.hasNextNode()) {
			beginNode = null;
			endNode = null;
		}
		else {
			beginNode = beginNode.getRefNextNodeOrNull();
		}
		
		elementCount--;
	}
	
	//method
	private void supposeIsNotEmpty() {
		if (isEmpty()) {
			throw new RuntimeException("The current list is empty.");
		}
	}
}
